### ISSUE TITLE
Possible to auto add hosts that do not exist in the known_hosts file

### ISSUE THREAD

**Issue by xNinjaKittyx** [ronf/asyncssh#176]
Is it possible in asyncssh to do something like how paramiko has `AutoAddPolicy`, `MissingHostKeyPolicy`, and `RejectPolicy`?

I believe the `RejectPolicy` is identical to the current default behavior, and `AutoAddPolicy` would almost essentially be the same as setting `known_hosts=None` (assuming you never use ssh outside of your python scripts...) .

I would like to be able to do something similar to `MissingHostKeyPolicy`

**Comment by ronf** [comment]
Are you looking for AsyncSSH to return an exception when there's no matching host key, or are you looking to provide your own callback? If you want an exception, that would generally be its default behavior. You should get back an asyncssh DisconnectError with a reason string of "No trusted server host keys available".

One thing that can get in the way of this is if you have any trusted X.509 certificates installed. By default, AsyncSSH will look in $HOME/.ssh/ca-bundle.crt and the directory $HOME/.ssh/crt. If it finds certificates in either of those places and doesn't find any other trusted server keys/certificates in the known hosts file, it will ask the server to use an X.509 based host key exchange, which most SSH servers don't support. This can cause the server to drop the connection, resulting in a DisconnectError with a reason like "Connection reset by peer" instead of "No trusted server host keys available".

I'm not sure offhand what exception paramiko raises when RejectPolicy is used, but I think you're right that this would be a close match to the default AsyncSSH behavior. There is no exact equivalent to paramiko's AutoAddPolicy right now, but setting known_hosts to None would disable the validation of the server host key, so you wouldn't need to have any server host keys added to AsyncSSH for it to be able to open the connection, even when no existing known hosts entry was present yet.

If you're looking for a callback that passes you the key, that's not something which is supported today but I could look into adding this as an optional method on the SSHClient class, similar to the callbacks there now for handling client-side authentication. I'm thinking of something like validate_server_public_key() if the server presents a host key and validate_server_ca_key() if the server presents a certificate. The caller would be expected to do their own validation of the host key or CA key passed as an argument, and return a boolean of whether to accept it or not. I could also potentially allow an exception to be raised here that could be passed back to the caller in place of the existing DisconnectError of "No trusted server host keys available".

**Comment by xNinjaKittyx** [comment]
I think what I meant to say is if there's a way to automatically add the key to the `known_hosts` file when the key is missing, which is what `MissingHostKeyPolicy` does for Paramiko, which could possibly also require the callback with the key? I was hoping this would be done internally, rather than requiring the user to do it manually.

**Comment by ronf** [comment]
I'm no expert on paramiko's behavior, but from looking at the documentation, MissingHostKey just provides a callback. It is the AutoAddPolicy which adds unknown host keys automatically, but the documentation says it adds them to the "local HostKeys" object. I'm guessing that means this key is only added in memory to whatever the host keys object associated with the SSHClient object was used to initiate the SSH connection, rather than being written back to something like "~/.ssh/known_hosts". That makes sense to me, but there's really no equivalent to that right now in AsyncSSH, as a new AsyncSSH SSHClient object is created for every new connection that's opened. So, there's no object shared across connections that could be used to hold previously seen host keys. Such an object would have to be created by the caller, and somehow made available through something like the client_factory passed into AsyncSSH's create_connection() call.

If you're thinking this could be done via the object passed in as the known_host argument, keep in mind that this argument can take many different forms, ranging from a filename or byte array in OpenSSH known hosts format to a callable function which will return a set of trusted and revoked keys and certificates which match a given host, address, and port.

I suppose it would be possible to build a subclass of SSHKnownHosts which starts out empty and just adds whatever is passed into it to the trusted list, keyed by the host, address, and port being matched. An instance of this could be instantiated by the caller and then passed to all of the connect calls that wanted to share a set of automatically trusted host keys. The security of this is only mildly better than passing in ``None``, though, as basically it would trust whatever key each host initially returned.

Also, this would be a bit dangerous -- once a host key was trusted, the lookup would fail if that host ever changed keys. Would you need some function on this object to forget about a particular host key once it was no longer being used so that a new host key could be learned? I presume you wouldn't want to automatically start trusting the new key, as that really WOULD be the same as passing in ``None`` and doing no check at all!

Some other questions also come to mind here:

- Would you ever want to match on address instead of hostname when deciding which key to trust? What if your lookup returns different keys when you look up by host vs. address? Should both be considered trusted for the new request?

- Should there be support for certificates, and if so do you trust all certificates signed by whatever CA you saw the first time you contacted the host, or do you still only trust a specific host key found in the certificate and ignore the CA which signed the certificate?

**Comment by xNinjaKittyx** [comment]
> Would you need some function on this object to forget about a particular host key once it was no longer being used so that a new host key could be learned? 

If the user decides to use a MissingHostKey Policy like behavior, it would be the caller's responsibility to remove the host key so that a new one could be added. Now that I think about it, I can't really think of a good reason to use MissingHostKey, since most of the time that I use AutoAddPolicy is within a company's lab network, mostly for convenience.

> I presume you wouldn't want to automatically start trusting the new key, as that really WOULD be the same as passing in None and doing no check at all!

That's essentially what paramiko's AutoAddPolicy does, but again, this is just an available feature for the caller, and not something that should be set by default. It wouldn't technically be exactly the same as using `None` since you could add the new keys to known_hosts, which would be utilized outside of running python. But obviously it doesn't mean anything for people who only run ssh through asynchssh and other ssh libraries that do similar behavior.

As for the last 2 questions, I'm not an expert on SSH behavior, so I wouldn't be the right person to give an answer.

**Comment by ronf** [comment]
> >  I presume you wouldn't want to automatically start trusting the new key, as that really WOULD be the same as passing in None and doing no check at all!

> That's essentially what paramiko's AutoAddPolicy does, but again, this is just an available feature for the caller, and not something that should be set by default. It wouldn't technically be exactly the same as using None since you could add the new keys to known_hosts, which would be utilized outside of running python. But obviously it doesn't mean anything for people who only run ssh through asynchssh and other ssh libraries that do similar behavior.

Does paramiko's AudioAddPolicy actually add keys to a file? I've never used it, but that's not at all the impression I got from reading the documentation. It looked to me like it only added keys in memory, to a specific SSHClient's list of trusted keys, meaning these keys were only trusted for requests using that specific SSHClient and the keys would be thrown away when that python session ended (or possibly even sooner if the SSHClient ceased to be used).

If you mean that the paramiko caller could add keys to ~/.ssh/known_hosts, that would require them to implement their own MissingHostKey subclass with that behavior rather than using AutoAddPolicy. If anything, some flavor of that is what I'm leaning toward here. You could implement your own AutoAddPolicy equivalent if you really wanted to with that callback, but it would be up to you to manage things like when keys got deleted from the automatic list to allow for adding a new key for a host. There's really no way anything I provided could do that automatically (at least not without disabling validation entirely, which is already available by passing in None).

Without knowing when host keys are changed, I just don't see how to make an AutoAddPolicy that's reliable.

**Comment by xNinjaKittyx** [comment]
I'm not entirely sure either if it adds the keys, so I'll have to look more into it.

I partially understand this callback that it would allow me to receive the key.

How would the workflow/code look like after receiving the key from a callback? I'm still fairly new to asyncssh.

**Comment by ronf** [comment]
The first thought I had was to add two new callbacks to the SSHClient class. They would look something like:

    validate_server_public_key(self, key)

and:

    validate_server_ca_key(self, key)

They would be called in the case where the known_hosts in the call to create_connection() didn't match anything. If the server provided a public key, validate_server_public_key() would be called with that key. If the server provided a certificate, valid_server_ca_key() would be called with the CA key used to sign the certificate.

These callbacks would be able to return True or False to indicate whether to trust the provided key or not. If they returned True, that would be considered a known_hosts match. In the case of a certificate, additional validation would then be done to make sure the certificate is valid, just as happens now when there's a known hosts match on the CA. I'd probably also support these callbacks raising a ValueError to specify a custom reason for why a key wasn't trusted. This is similar to the way the paramiko MissingHostKey() callback works.

If you wanted to implement something like AutoAddPolicy(), you could keep your own list of previously seen keys in a custom SSHClient object, adding keys to the list if you've never seen any key for a given host, but after that only trusting the remembered key for that host. It would be up to you to decide when keys should be removed from that list to allow a new key to be trusted, and also to decide how you associated keys with the host/address/port you were connecting to when looking up whether you had a previously seen key or not.

**Comment by ronf** [comment]
One other thing to be aware of -- the matching of known_hosts is used to determine what to client sends as trusted host key algorithms to the server, and this happens _before_ the server sends its host key. In fact, it is used by the server to decide which host key to send. With no matching host keys known in advance, this won't be possible.

This issue already exists when known_hosts is set to None, and in that case AsyncSSH just sends a list of _all_ host key algorithms it supports. However, I just wanted to mention this as another point to consider. Unlike other cases where there are arguments to specify which algorithms are allowed, that doesn't exist here, since the expectation was that the filtering could be done by limiting what specific host/CA keys were trusted.

**Comment by ronf** [comment]
Ok - I've implemented a first cut of this in commit 8c7e170 in the develop branch. Documentation can be found at https://asyncssh.readthedocs.io/en/develop/api.html#sshclient and https://asyncssh.readthedocs.io/en/develop/api.html#sshserver.

The final APIs ended up looking like the following in SSHClient:

```python
    validate_host_public_key(self, host, addr, port, key)

    validate_host_ca_key(self, host, addr, port, key)
```

I added a similar capability to support dynamic validation of client host keys in SSHServer:

```python
    validate_host_public_key(self, client_host, client_addr, client_port, key)

    validate_host_ca_key(self, client_host, client_addr, client_port, key)
```

In addition, if no known_client_hosts are provided in advance, the SSHServer subclass will need to implement host_based_auth_supported() and have it return True for host-based authentication to be enabled.

If you wanted to implement something like AutoAddPolicy, you'd just need to have these callbacks keep their own dictionary of what keys have been seen before for what hosts. If there's no entry for a host, it would be added to the dictionary and True would be returned. Otherwise, if there is already a key but it doesn't match what was passed in, False should be returned to reject the key.

You could optionally support matching against the address as well, and decide whether or not you wanted to match on port (to allow different keys to be used if multiple SSH servers were running on the same host on different ports).

**Comment by xNinjaKittyx** [comment]
Thanks for the update.

So, would I have create a child class of `SSHClient` that would implement a `validate_host_public_key` to keep a dictionary internally, correct?

When `known_hosts` is not `None` when using `create_connection()` does it ignore this callback?

**Comment by ronf** [comment]
Yes, that's right - that and/or validate_host_ca_key() if you want to be able to support OpenSSH certificates used as server host keys, where you match against the certificate authority key used to sign the certificate.

If known hosts are specified, they will be tested against before calling these new callbacks. The callbacks will only be called when there's no match against the provided known_hosts.

**Comment by ronf** [comment]
The support described here is now released in AsyncSSH 1.15.1.
